File: go/bob/bob.go
package bob

import (
	"regexp"
	"strings"
)

func say(s string) bool {
	return len(s) == 0
}

func yell(s string) bool {
	re := regexp.MustCompile("[A-Za-z]+")
	ws := re.FindAllString(s, -1)
	y := len(ws) != 0
	for _, w := range ws {
		y = y && strings.ToUpper(w) == w
	}
	return y
}

func ask(s string) bool {
	return strings.HasSuffix(s, "?")
}

// Hey bob
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)

	switch {
	case say(remark):
		return "Fine. Be that way!"
	case yell(remark):
		return "Whoa, chill out!"
	case ask(remark):
		return "Sure."
	default:
		return "Whatever."
	}
}

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      say?(input)  -> "Fine. Be that way!"
      ask?(input)  -> "Sure."
      yell?(input) -> "Whoa, chill out!"
      true         -> "Whatever."
    end
  end

  defp say?(input) do
    String.trim(input) == ""
  end

  defp ask?(input) do
    String.ends_with?(input, "?")
  end

  defp yell?(input) do
    input == String.upcase(input) && input != String.downcase(input)
  end
end
